libraries/chain/controller.cpp

```diff
@@ -1212,23 +1218,28 @@ struct controller_impl {
                      ("lib_num", lib_num)("bn", fork_db_root_block_num()) );
       }
 
-      const uint32_t if_lib = if_irreversible_block_num;
+      uint32_t if_lib = block_header::num_from_id(if_irreversible_block_id);
       const uint32_t new_lib = if_lib > 0 ? if_lib : fork_db_head_irreversible_blocknum();
 
       if( new_lib <= lib_num )
          return;
 
       auto mark_branch_irreversible = [&, this](auto& forkdb) {
-         auto branch = forkdb.fetch_branch( fork_db_head(forkdb, irreversible_mode())->id(), new_lib );
+         auto branch = (if_lib > 0) ? forkdb.fetch_branch( if_irreversible_block_id, new_lib)
```

Comment:
That's always going to be a single block, right, since `new_lib == num_from_id(if_irreversible_block_id);`

Reply:
No, this returns all blocks from root up to `new_lib` on branch with `if_irreversible_block_id`.

Reply:
Aren't they the same though?

Reply:
Looking at the code I see:
```
if_lib = block_header::num_from_id(if_irreversible_block_id);
new_lib = if_lib;
```

Reply:
`new_lib` does equal `num_from_id(if_irreversible_block_id)`

However, `forkdb.fetch_branch( if_irreversible_block_id, new_lib)` returns all blocks from the `root` of the fork database up to and including `if_irreversible_block_id` trimming off any after `new_lib` which there will not be any. Could not pass in `new_lib` as it doesn't matter. However, might as well pass it in as it makes it same as pre-instant-finality.

Here is an example from a unittest where LIB moved by two blocks. You can see the branch is of size 2 in this case.

![image](https://github.com/AntelopeIO/leap/assets/3933649/fa6004c8-9a92-4876-ad3c-0c38f0b1b2bf)

Reply:
Got it, sorry, for some reason I keep misunderstanding that `fetch_branch` API.